politeiavoter/politeiavoter.go

```diff
@@ -615,6 +658,51 @@ func (c *ctx) _vote(token, voteId string) ([]string, *v1.BallotReply, error) {
 		return nil, nil, fmt.Errorf("no eligible tickets found")
 	}
 
+	// Put cast votes in a map so that we can
+	// filter in linear time
+	castVotes := make(map[string]v1.CastVote)
+	for _, v := range vrr.CastVotes {
+		castVotes[v.Ticket] = v
+	}
+
+	// Filter out tickets that have already voted
+	filtered := make([]*pb.CommittedTicketsResponse_TicketAddress, 0,
+		len(ctres.TicketAddresses))
+	for _, t := range ctres.TicketAddresses {
+		h, err := chainhash.NewHash(t.Ticket)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		v, ok := castVotes[h.String()]
+		if ok {
+			// Ticket has already voted. Validate signature.
+			sig, err := hex.DecodeString(v.Signature)
```

Comment:
I'd suggest making a `func verifyV1Vote(address string, vote *v1.CastVote) bool` function which handles converting the signature, creating the message, and verifying it in order simplify all this.

The result would simply be:
```Go
for _, t := range ctres.TicketAddresses {
    v, ok := castVotes[t.Ticket]
    if !ok || !verifyV1Vote(t.Address, v) {
        filtered = append(filtered, t)
    }
}
```

Reply:
I agree. This has been updated.